src/api/tap.rs

```diff
@@ -0,0 +1,322 @@
+async fn tap_handler(
+    patterns: HashSet<String>,
+    tx: TapSender,
+    mut watch_rx: WatchRx,
+    mut shutdown_rx: ShutdownRx,
+) {
+    debug!(message = "Started tap.", patterns = ?patterns);
+
+    // Sinks register for the current tap. Will be updated as new components match.
+    let mut sinks = HashMap::new();
+
+    // Keep a copy of the last topology snapshot, for later clean-up.
+    let mut last_outputs = None;
+
+    loop {
+        tokio::select! {
+            _ = &mut shutdown_rx => break,
+            Some(outputs) = watch_rx.recv() => {
+                // Get the patterns that matched on the last iteration, to compare with the latest
+                // round of matches when sending notifications.
+                let last_matches = patterns
+                    .iter()
+                    .filter(|pattern| sinks.keys().any(|name: &String| pattern.matches_glob(name)))
+                    .collect::<HashSet<_>>();
+
+                // Iterate over outputs, returning a set of matched patterns from this latest round.
+                let matched = outputs
+                    .iter()
+                    .filter_map(|(name, control_tx)| {
+                        match patterns
+                            .iter()
+                            .filter(|pattern| pattern.matches_glob(name))
+                            .collect_vec()
+                        {
+                            matched if !matched.is_empty() => {
+                                debug!(
+                                    message="Component matched.",
+                                    component_name = ?name, patterns = ?patterns, matched = ?matched
+                                );
+
+                                // (Re)connect the sink. This is necessary because a sink may be
+                                // reconfigured with the same name as a previous, and we are not
+                                // getting involved in config diffing at this point.
+                                let id = Uuid::new_v4().to_string();
+                                let sink = make_router(tx.clone(), name);
+                                if control_tx
+                                    .send(fanout::ControlMessage::Add(id.to_string(), sink))
+                                    .is_ok()
+                                {
+                                    // (Over)write the sink entry.
+                                    debug!(
+                                        message = "Component connected.",
+                                        component_name = ?name, id = ?id
+                                    );
+                                    sinks.insert(name.to_string(), id);
+                                } else {
+                                    error!(
+                                        message = "Couldn't connect component.",
+                                        component_name = ?name, id = ?id
```

Comment:
Should we output the error here?

Reply:
By output, do you mean surface this to the client? We certainly could. Notifications aren't currently accommodating to error states, but there's no reason they couldn't be expanded. In practice, this shouldn't happen and would indicate something has gone wrong upstream, which felt like more a Vector-level error than something the client could do anything about, but I could see an argument for surfacing that too.

Reply:
Ah, sorry, I just mean include it in the error message here.

Reply:
Updated in c2d335ed2190e2df2dfa0f9763bd19f172a9311b.

This isn't especially interesting; the control channel here is of type `tokio::mpsc::UnboundedSender<ControlMessage>`, which returns a `Result<(), SendError<T>>`.

SendError's Debug implementation returns a static string, because there can only be one error condition:

```rust
#[derive(Debug)]
pub struct SendError<T>(pub T);

impl<T> fmt::Display for SendError<T> {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(fmt, "channel closed")
    }
}
``` 

But it's at least a bit more context for anyone trawling logs.